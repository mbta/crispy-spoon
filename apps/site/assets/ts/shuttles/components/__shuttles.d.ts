import { DirectionId } from "../../__v3api";

export type MaybeDirectionId = DirectionId | null;

export interface Stop {
  id: string;
  direction_id: MaybeDirectionId;
  headsign: string | null;
  longitude: number;
  latitude: number;
  name: string;
  type: "rail_affected" | "rail_unaffected" | "shuttle";
}

export interface Shape {
  color: string;
  direction_id: MaybeDirectionId;
  id: string;
  is_shuttle_route: boolean;
  polyline: [number, number][];
}

export interface Diversion {
  shapes: Shape[];
  stops: Stop[];
}
