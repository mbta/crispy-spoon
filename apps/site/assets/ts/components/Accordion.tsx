import React, { ReactElement, useState } from "react";

// A simple accordion implementation, based on site_web/templates/partial/_accordion_ui
// This attempts to mimic the usage of <ExpandableBlock />, where the contents rerender with expanded state
interface Props {
  id: string;
  title: {
    collapsed: string;
    expanded: string;
  };
  children: ReactElement<HTMLElement>;
}

export const AccordionNoJS = ({
  id,
  children
}: {
  id: string;
  children: ReactElement<HTMLElement>;
}): ReactElement<HTMLElement> => (
  <div className="c-accordion-ui">
    <div
      className="c-accordion-ui__target"
      id={`${id}-section`}
      role="region"
      aria-labelledby={`${id}-title`}
    >
      <div className="c-accordion-ui__content">{children}</div>
    </div>
  </div>
);

const Accordion = (props: Props): ReactElement<HTMLElement> => {
  const { id, title, children } = props;
  const [isExpanded, toggleExpanded] = useState(false);
  const onClick = (): void => toggleExpanded(expanded => !expanded);
  return (
    <div className="c-accordion-ui">
      <div className="panel">
        <div className="c-accordion-ui__heading" role="heading" aria-level={3}>
          <button
            className={`c-accordion-ui__trigger ${!isExpanded && "collapsed"}`}
            aria-expanded={isExpanded}
            aria-controls={`${id}-section`}
            onClick={onClick}
            type="button"
          >
            <span className="c-accordion-ui__title" id={`${id}-title`}>
              {isExpanded ? title.expanded : title.collapsed}
            </span>
            <div className="c-accordion-ui__indicator">
              <span className="c-indicator__content c-indicator__content--angle" />
            </div>
          </button>
        </div>
        <div
          className="c-accordion-ui__target js-focus-on-expand"
          id={`${id}-section`}
          role="region"
          aria-labelledby={`${id}-title`}
        >
          {isExpanded && (
            <div className="c-accordion-ui__content">{children}</div>
          )}
        </div>
      </div>
    </div>
  );
};

export default Accordion;
