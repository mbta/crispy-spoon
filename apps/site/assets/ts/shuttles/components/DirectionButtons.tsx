import React, { ReactElement, useState, useEffect } from "react";
import { MaybeDirectionId } from "./__shuttles";
import { DirectionInfo } from "../../__v3api";

interface Props {
  places: DirectionInfo;
  onClick: (directionId: MaybeDirectionId) => void;
}

const DirectionButtons: React.FC<Props> = ({
  places,
  onClick
}: Props): ReactElement<HTMLElement> => {
  const [selectedDirectionId, setSelectedDirectionId] = useState<
    MaybeDirectionId
  >(null);

  useEffect(
    () => {
      onClick(selectedDirectionId);
    },
    [onClick, selectedDirectionId]
  );

  const DirectionButton = (
    directionId: MaybeDirectionId,
    headsign: string
  ): JSX.Element => (
    <button
      key={`${headsign}`}
      type="button"
      className={`btn btn-secondary btn-sm c-btn-group-stackable__btn ${selectedDirectionId ===
        directionId && "c-btn-group-stackable__btn--active"}`}
      onClick={() => {
        setSelectedDirectionId(directionId);
      }}
    >
      {headsign}
    </button>
  );

  return (
    <div
      className="c-btn-group-stackable"
      role="group"
      aria-label="Filter shuttle map by direction"
    >
      {DirectionButton(null, "All Directions")}
      {Object.entries(places)
        .sort()
        .map(([directionId, headsign]) =>
          DirectionButton(
            parseInt(directionId, 10) as MaybeDirectionId,
            headsign
          )
        )}
    </div>
  );
};

export default DirectionButtons;
