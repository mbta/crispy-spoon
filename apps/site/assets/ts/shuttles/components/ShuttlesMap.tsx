import React, { ReactElement } from "react";
import { Stop, Shape } from "./__shuttles";
import Map from "../../leaflet/components/Map";
import {
  MapData,
  MapMarker,
  Polyline,
  IconOpts,
  TileServerUrl
} from "../../leaflet/components/__mapdata";

interface Props {
  shapes: Shape[];
  stops: Stop[];
  centerStop?: Stop;
  tileServerUrl: TileServerUrl;
}

/* eslint-disable @typescript-eslint/camelcase */
const makeMarkers = (stops: Stop[]): MapMarker[] => {
  const iconMap = {
    shuttle: "mode-bus-default",
    rail_affected: "circle-t-small-grey",
    rail_unaffected: "circle-t-small"
  };
  const icon_opts: IconOpts = {
    icon_size: [16, 16],
    icon_anchor: [8, 8],
    popup_anchor: [0, -8]
  };
  const rotation_angle = 0;

  return stops.map((stop: Stop, index) => {
    const icon = iconMap[stop.type] || "circle-t-small";
    const id = `${stop.name}-${index}`;
    const { longitude, latitude } = stop;
    const tooltip = (
      <span>
        {stop.name}
        {stop.type === "shuttle" && ` Shuttle Pickup (${stop.headsign})`}
      </span>
    );

    return {
      icon,
      icon_opts,
      id,
      latitude,
      longitude,
      rotation_angle,
      tooltip,
      shape_id: stop.id
    };
  });
};
/* eslint-enable @typescript-eslint/camelcase */

const makePolylines = (shapes: Shape[]): Polyline[] =>
  shapes.map((shape: Shape, index) => ({
    color: !shape.is_shuttle_route && shape.color ? `#${shape.color}` : "black",
    "dotted?": shape.is_shuttle_route,
    id: `${shape.id}-${index}`,
    positions: shape.polyline,
    weight: shape.is_shuttle_route ? 4 : 6
  }));

const ShuttlesMap = ({
  shapes,
  stops,
  centerStop,
  tileServerUrl
}: Props): ReactElement<HTMLElement> | null => {
  if (typeof window !== "undefined") {
    /* eslint-disable-next-line global-require */
    const getBounds = require("../../leaflet/bounds").default;
    const defaultBounds = getBounds(
      makeMarkers(stops.filter(stop => stop.type === "rail_affected"))
    );

    const bounds = centerStop
      ? getBounds(makeMarkers([centerStop]))
      : defaultBounds;

    // the MapData interface requires a center value
    const { lng: longitude, lat: latitude } = defaultBounds.getCenter();
    const defaultCenter = centerStop
      ? {
          longitude: centerStop.longitude,
          latitude: centerStop.latitude
        }
      : { longitude, latitude };

    /* eslint-disable @typescript-eslint/camelcase */
    const mapData: MapData = {
      height: 400,
      width: 400,
      default_center: defaultCenter,
      markers: makeMarkers(stops),
      polylines: makePolylines(shapes),
      tile_server_url: tileServerUrl,
      zoom: centerStop ? 18 : 13
    };
    /* eslint-enable @typescript-eslint/camelcase */

    return (
      <div className="m-shuttles-map">
        <Map mapData={mapData} bounds={bounds} boundsByMarkers={false} />
      </div>
    );
  }
  return null;
};

export default ShuttlesMap;
