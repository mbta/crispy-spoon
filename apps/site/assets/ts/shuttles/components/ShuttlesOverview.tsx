import React, { ReactElement, useState, useEffect } from "react";

interface Props {
  places: {
    0: string;
    1: string;
  };
}

const ShuttlesOverview = ({ places }: Props): ReactElement<HTMLElement> => {
  const [selectedDirectionId, setSelectedDirectionId] = useState("");
  useEffect(() => {}, [selectedDirectionId]); // placeholder function, update map here

  const DirectionButton = (
    directionId: string,
    headsign: string
  ): JSX.Element => (
    <button
      key={`${headsign}`}
      type="button"
      className={`btn btn-secondary btn-sm ${
        selectedDirectionId === directionId
          ? "c-btn-group-stackable__btn c-btn-group-stackable__btn--active"
          : "c-btn-group-stackable__btn"
      }`}
      onClick={() => setSelectedDirectionId(directionId)}
    >
      {headsign}
    </button>
  );

  return (
    <>
      <div
        className="c-btn-group-stackable"
        role="group"
        aria-label="Filter shuttle map by direction"
      >
        {DirectionButton("", "All Directions")}
        {Object.entries(places)
          .sort()
          .map(([directionId, headsign]) =>
            DirectionButton(directionId, headsign)
          )}
      </div>
      <div>Map1 {selectedDirectionId} </div>
    </>
  );
};

export default ShuttlesOverview;
