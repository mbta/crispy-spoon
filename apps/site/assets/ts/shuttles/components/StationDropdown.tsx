import React, { ReactElement, useState, useEffect } from "react";
import renderSvg from "../../helpers/render-svg";
import arrowIcon from "../../../static/images/icon-down-arrow.svg";
import { Stop } from "./__shuttles";

interface Props {
  stations: Stop[];
  onSelect: (stationId: string) => void;
}

const StationDropdown = ({
  stations,
  onSelect
}: Props): ReactElement<HTMLElement> => {
  const [selectedStationId, setSelectedStationId] = useState(stations[0].id);
  useEffect(
    () => {
      onSelect(selectedStationId);
    },
    [onSelect, selectedStationId]
  );

  return (
    <div className="c-select-custom__container">
      <select
        className="c-select-custom"
        defaultValue={selectedStationId}
        onChange={e => setSelectedStationId(e.target.value)}
      >
        {stations.map(station => (
          <option key={station.id} value={station.id}>
            {station.name}
          </option>
        ))}
      </select>
      {renderSvg("c-svg__icon c-select-custom__arrow", arrowIcon)}
    </div>
  );
};

export default StationDropdown;
