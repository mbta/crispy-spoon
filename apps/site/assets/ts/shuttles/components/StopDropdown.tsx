import React, {
  ReactElement,
  useState,
  useEffect,
  SetStateAction,
  Dispatch
} from "react";
import renderSvg from "../../helpers/render-svg";
import arrowIcon from "../../../static/images/icon-down-arrow.svg";
import { Stop } from "./__shuttles";

interface Props {
  stops: Stop[];
  onSelect: Dispatch<SetStateAction<Stop>>;
}

const StopDropdown = ({
  stops,
  onSelect
}: Props): ReactElement<HTMLElement> => {
  const [selectedStopId, setSelectedStopId] = useState(stops[0].id);
  useEffect(
    () => {
      const selectedStop = stops.find(stop => stop.id === selectedStopId);
      if (selectedStop) {
        onSelect(selectedStop);
      }
    },
    [onSelect, selectedStopId, stops]
  );

  return (
    <div className="c-select-custom__container">
      <select
        className="c-select-custom"
        value={selectedStopId}
        onChange={e => setSelectedStopId(e.target.value)}
      >
        {stops.map(stop => (
          <option key={stop.id} value={stop.id}>
            {stop.name}
          </option>
        ))}
      </select>
      {renderSvg("c-svg__icon c-select-custom__arrow", arrowIcon)}
    </div>
  );
};

export default StopDropdown;
