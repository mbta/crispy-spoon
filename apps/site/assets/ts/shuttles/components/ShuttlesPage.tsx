import React, { ReactElement, useState, useEffect } from "react";
import ShuttlesMap from "./ShuttlesMap";
import StopDropdown from "./StopDropdown";
import { Route } from "../../__v3api";
import { Diversion, MaybeDirectionId } from "./__shuttles";
import { TileServerUrl } from "../../leaflet/components/__mapdata";

interface Props {
  route: Route;
  diversions: Diversion;
  tileServerUrl: TileServerUrl;
}

const ShuttlesPage: React.FC<Props> = ({
  route,
  diversions,
  tileServerUrl
}: Props): ReactElement<HTMLElement> => {
  const places =
    route.id === "Green" ? route.direction_names : route.direction_destinations;
  const initialShapes = [...diversions.shapes].map(shape => ({
    color: route.color,
    ...shape
  }));
  const initialStops = [...diversions.stops].map(stop => ({
    headsign:
      stop.direction_id === null ? "All Directions" : places[stop.direction_id],
    ...stop
  }));
  const [selectedDirectionId, setSelectedDirectionId] = useState<
    MaybeDirectionId
  >(null);
  const [displayedShapes, setDisplayedShapes] = useState(initialShapes);
  const [displayedStops, setDisplayedStops] = useState(initialStops);
  const [selectedStop, setSelectedStop] = useState(
    initialStops.sort().filter(s => s.type === "rail_affected")[0]
  );

  const filterMap = (): void => {
    if (selectedDirectionId === null) {
      setDisplayedShapes(initialShapes);
      setDisplayedStops(initialStops);
    } else {
      setDisplayedShapes(
        initialShapes.filter(
          shape => shape.direction_id === selectedDirectionId
        )
      );
      setDisplayedStops(
        initialStops.filter(
          stop =>
            stop.direction_id === selectedDirectionId ||
            stop.direction_id === null
        )
      );
    }
  };

  useEffect(filterMap, [selectedDirectionId]);

  const DirectionButton = (
    directionId: MaybeDirectionId,
    headsign: string
  ): JSX.Element => (
    <button
      key={`${headsign}`}
      type="button"
      className={`btn btn-secondary btn-sm c-btn-group-stackable__btn ${selectedDirectionId ===
        directionId && "c-btn-group-stackable__btn--active"}`}
      onClick={() => setSelectedDirectionId(directionId)}
    >
      {headsign}
    </button>
  );

  return (
    <div className="shuttles__main">
      <h2>Shuttle Maps</h2>

      <div
        className="c-btn-group-stackable"
        role="group"
        aria-label="Filter shuttle map by direction"
      >
        {DirectionButton(null, "All Directions")}
        {Object.entries(places)
          .sort()
          .map(([directionId, headsign]) =>
            DirectionButton(
              parseInt(directionId, 10) as MaybeDirectionId,
              headsign
            )
          )}
      </div>

      <ShuttlesMap
        tileServerUrl={tileServerUrl}
        shapes={displayedShapes}
        stops={displayedStops}
      />

      <div className="c-heading-with-control">
        <h3>Stop Detail</h3>

        <StopDropdown
          stops={diversions.stops
            .sort()
            .filter(stop => stop.type === "rail_affected")}
          onSelect={setSelectedStop}
        />
      </div>
      <ShuttlesMap
        tileServerUrl={tileServerUrl}
        shapes={displayedShapes}
        stops={displayedStops}
        centerStop={selectedStop}
      />
    </div>
  );
};

export default ShuttlesPage;
