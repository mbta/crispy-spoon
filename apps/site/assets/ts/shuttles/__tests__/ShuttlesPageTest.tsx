import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import _diversionsData from "./diversionData.json";
import ShuttlesPage from "../components/ShuttlesPage";
import { Route } from "../../__v3api";
import { Diversion } from "../components/__shuttles";
import { TileServerUrl } from "../../leaflet/components/__mapdata";
import ShuttlesMap from "../components/ShuttlesMap";
import StopDropdown from "../components/StopDropdown";
import renderer from "react-test-renderer";

const diversionsData = _diversionsData as Diversion;
const tileServerUrl: TileServerUrl =
  "https://mbta-map-tiles-dev.s3.amazonaws.com";

const route: Route = {
  color: "red",
  description: "",
  direction_destinations: { 0: "Alewife", 1: "Ashmont / Braintree" },
  direction_names: { 0: "Inbound", 1: "Outbound" },
  id: "Red",
  name: "Red",
  long_name: "Red Line",
  type: 1
};

const greenRoute: Route = {
  color: "green",
  description: "",
  direction_names: { "0": "Westbound", "1": "Eastbound" },
  direction_destinations: {
    "0": "Boston College / Cleveland Circle / Riverside / Heath Street",
    "1": "Park Street / Government Center / North Station / Lechmere"
  },
  id: "Green",
  name: "Green",
  long_name: "Green Line",
  type: 0
};

const names = (r: Route) => Object.values(r.direction_names);
const destinations = (r: Route) => Object.values(r.direction_destinations);

jest.mock("react-leaflet/lib/Map");
let wrapper: ShallowWrapper;
let greenWrapper: ShallowWrapper;

/* use shallow instead of mount. mount throws Uncaught [RangeError: Maximum call stack size exceeded] 
from a dependency, fast-deep-equal, used to memoize the Map component ShuttlesMap is based on. */
beforeEach(() => {
  wrapper = shallow(
    <ShuttlesPage
      route={route}
      diversions={diversionsData}
      tileServerUrl={tileServerUrl}
    />
  );
  greenWrapper = shallow(
    <ShuttlesPage
      route={greenRoute}
      diversions={diversionsData}
      tileServerUrl={tileServerUrl}
    />
  );
});

afterEach(() => {
  wrapper!.unmount();
  greenWrapper!.unmount();
});

it("the shuttles page renders", () => {
  const tree = renderer
    .create(
      <ShuttlesPage
        route={route}
        diversions={diversionsData}
        tileServerUrl={tileServerUrl}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("the shuttles page shows direction names for combined 'Green' Line", () => {
  destinations(greenRoute).forEach(headsign => {
    expect(
      greenWrapper.containsMatchingElement(<button>{headsign}</button>)
    ).toEqual(false);
  });
  names(greenRoute).forEach(name => {
    expect(
      greenWrapper.containsMatchingElement(<button>{name}</button>)
    ).toEqual(true);
  });
});

it("the shuttles page shows direction destinations for all other lines", () => {
  destinations(route).forEach(headsign => {
    expect(
      wrapper.containsMatchingElement(<button>{headsign}</button>)
    ).toEqual(true);
  });
  names(route).forEach(name => {
    expect(wrapper.containsMatchingElement(<button>{name}</button>)).toEqual(
      false
    );
  });
});

it("the shuttles page by default shows all stops", () => {
  wrapper.find(ShuttlesMap).forEach(
    (node: any): void => {
      const numStops = node.prop("stops").length;
      expect(numStops).toBe(diversionsData.stops.length);
    }
  );
});

it("the shuttles page by default shows all shapes", () => {
  wrapper.find(ShuttlesMap).forEach(
    (node: any): void => {
      const numShapes = node.prop("shapes").length;
      expect(numShapes).toBe(diversionsData.shapes.length);
    }
  );
});

it.each`
  index | direction | name
  ${0}  | ${null}   | ${"All Directions"}
  ${1}  | ${0}      | ${route.direction_destinations["0"]}
  ${2}  | ${1}      | ${route.direction_destinations["1"]}
`(
  "the shuttles page button for direction $direction has name $name",
  ({ index, name }) => {
    expect(
      wrapper
        .find("button")
        .at(index)
        .text()
    ).toEqual(name);
  }
);

it("the shuttles page <select> lists affected stops only", () => {
  const options = wrapper
    .find(StopDropdown)
    .dive()
    .find("select option")
    .map((node: ShallowWrapper) => node.text());
  const affectedStops = diversionsData.stops
    .sort()
    .filter(s => s.type === "rail_affected")
    .map(s => s.name);
  expect(options).toEqual(affectedStops);
});

it("the shuttles page detail map by default centers first affected stop", () => {
  let map = wrapper.find(ShuttlesMap).at(1);
  const firstStation = diversionsData.stops
    .sort()
    .filter(s => s.type === "rail_affected")[0];
  expect(map.prop("centerStop")!.id).toEqual(firstStation.id);
});
