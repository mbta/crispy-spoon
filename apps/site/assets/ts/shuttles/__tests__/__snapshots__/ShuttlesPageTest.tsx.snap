// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the shuttles page renders 1`] = `
<div
  className="shuttles__main"
>
  <h2>
    Shuttle Maps
  </h2>
  <div
    aria-label="Filter shuttle map by direction"
    className="c-btn-group-stackable"
    role="group"
  >
    <button
      className="btn btn-secondary btn-sm c-btn-group-stackable__btn c-btn-group-stackable__btn--active"
      onClick={[Function]}
      type="button"
    >
      All Directions
    </button>
    <button
      className="btn btn-secondary btn-sm c-btn-group-stackable__btn false"
      onClick={[Function]}
      type="button"
    >
      Alewife
    </button>
    <button
      className="btn btn-secondary btn-sm c-btn-group-stackable__btn false"
      onClick={[Function]}
      type="button"
    >
      Ashmont / Braintree
    </button>
  </div>
  <div
    className="m-shuttles-map"
  />
  <div
    className="c-heading-with-control"
  >
    <h3>
      Stop Detail
    </h3>
    <div
      className="c-select-custom__container"
    >
      <select
        className="c-select-custom"
        onChange={[Function]}
        value="place-knncl"
      >
        <option
          value="place-knncl"
        >
          Kendall/MIT
        </option>
        <option
          value="place-chmnl"
        >
          Charles/MGH
        </option>
        <option
          value="place-pktrm"
        >
          Park Street
        </option>
        <option
          value="place-dwnxg"
        >
          Downtown Crossing
        </option>
        <option
          value="place-sstat"
        >
          South Station
        </option>
        <option
          value="place-brdwy"
        >
          Broadway
        </option>
      </select>
      <span
        aria-hidden="true"
        className="notranslate c-svg__icon c-select-custom__arrow"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </div>
  </div>
  <div
    className="m-shuttles-map"
  />
</div>
`;
